File: leetcode/144-binary-tree-preorder-traversal/preorder_traversal.py
""" 
LC 144: Binary Tree Preorder Traversal

Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

"""
class Solution:
    
    # recursive solution
    def preorder_traversal(self, root: Optional[TreeNode]) -> list[int]:
        res = []
        
        def preorder(node):
            if not node:
                return

            res.append[node.val]
            preorder(node.left)
            preorder(node.right)
        
        preorder(root)
        return res
    
    # iterative solution
    def preorder_traversal_iter(self, root: Optional[TreeNode]) -> List[int]:
        curr, stack = root, []
        res = []
        
        while curr or stack:
            if curr:
                res.append(curr.val)
                stack.append(curr.right)
                curr = curr.left
            else:
                curr = stack.pop()

File: leetcode/144-binary-tree-preorder-traversal/PreorderTraversal.java
// LC 144: Preorder Traversal

class PreorderTraversal {

    private void preorder(TreeNode node, List<Integer> arr) {
            if(node == null)
                return;
            
            arr.add(node.val);
            preorder(node.left, arr);
            preorder(node.right, arr);
    }

    public List<Integer> preorderTraversal(TreeNode root) {
        
        List<Integer> res = new ArrayList<>();

        preorder(root, res);
        return res;

    }
}
